Fix: Can't edit the page if Favorite Widgets are used in it (and experiment is enabled) [ED-6166]

YitzhakAvm opened the pull request.

Commit:
Fix: Can't edit the page if Favorite Widgets are used in it (and experiment is enabled) [ED-6166]

StyleShit commented on includes/base/controls-stack.php:
```diff
@@ -998,6 +1010,14 @@ final public function get_config() {
 			}
 		}
 
+		foreach ( $this->additional_config as $key => $value ) {
+			if ( isset( $this->config[ $key ] ) && is_array( $this->config[ $key ] ) && is_array( $value ) ) {
+				$this->config[ $key ] = array_merge( $this->config[ $key ], $value );
+			} else {
+				$this->config[ $key ] = $value;
+			}
+		}
+
```
Not sure if it fits here, but maybe you can use [wp_parse_args()](https://developer.wordpress.org/reference/functions/wp_parse_args/)?

ItzhakAvm replied:
Good instinct!
Just gave it a shot and it works :)

StyleShit replied:
🔥

Commit:
Fix: Review Fixes

matipojo reviewed (changes requested):
Please add tests

ItzhakAvm commented:
@matipojo 
It's only happen when pro activated. Since playwright has no pro installed, currently there's no way to test it.

matipojo reviewed (changes requested):
Please add unit tests

Commit:
Add Unit Tests


tests - second approach


tests - third approach


tests - forth approach


tests - fifth approach


tests - seventh approach


tests - eigth approach


tests - ninth approach

matipojo commented on tests/phpunit/elementor/includes/base/test-widget-base.php:
```diff
@@ -31,6 +31,31 @@ public function tearDown() {
+		$widget = new Mock_Widget( [ 'settings' => [], 'id' => '1' ], [] );
+
+		$widget->set_config( 'test', [ 'first' ] );
+		$widget->set_config( 'test', [ 'second' ] );
```
Please add a test that overrides a value in the array
```
$widget->set_config( 'test', [ 'first' => 1 ] );
$widget->set_config( 'test', [ 'first' => 2 ] );
```

ItzhakAvm replied:
The desired behavior is:
`$widget->set_config( 'categories', [ 'favorites' ] ); // [ 'basic', 'favorites' ]`

Which means there should be a merge, not an override.

matipojo replied:
Please add a test that ensure it doesn't override. In order to avoid a future overriding.

matipojo commented on includes/base/controls-stack.php:
```diff
@@ -90,6 +90,18 @@ abstract class Controls_Stack extends Base_Object {
+	 * The additional configuration.
+	 *
+	 * Holds additional configuration that has been set using the `set_config`
+	 * method and later merged into the primary config.
```
Explain why it needed to be merged later

ItzhakAvm replied:
Done.

matipojo reviewed (changes requested):
The test fails

Commit:
Progress


f
